lib/vector-common/src/internal_event/bytes_sent.rs

```diff
@@ -20,3 +22,35 @@ impl<'a> InternalEvent for BytesSent<'a> {
         Some("BytesSent")
     }
 }
+
+#[allow(clippy::module_name_repetitions)]
+pub struct RegisteredBytesSent {
+    pub protocol: Cow<'static, str>,
+}
```

Comment:
Hmmm, I have to admit, seeing it in practice, where it's going to be either `Registered<RegisteredBytesSent>` or `<RegisteredBytesSent as RegisterInternalEvent>::Handle`.... it feels pretty blah.

Is there any way we could boil this down to `Registered<BytesSent>` such that we implemented `RegisterInternalEvent` on `BytesSent` itself? If we consider the overlap between `BytesSent` and `RegisteredBytesSent`, could we have something like this instead:

```rust
impl BytesSent {
  pub fn from_protocol<P: AsRef<str>>(protocol: P) -> Self {
    Self {
      byte_size: 0,
      protocol: protocol.as_ref(),
    }
  }
}

// move impl of `RegisterInternalEvent` from `RegisteredBytesSent` to `BytesSent`

// some code actually trying to use this:

struct Data {
  bytes_sent: Registered<BytesSent>,
}

impl Data {
  pub fn new() -> Self {
    Self {
      bytes_sent: register!(BytesSent::from_protocol("grpc")),
    }
  }
}
```

Essentially, it seems likely that the existing struct, and its `Registered*` companion, will almost always have overlapping fields based on the fields used as labels, so it might not be the prettiest thing, but we could save a lot of duplication if we just piggybacked off of those existing fields.

Admittedly, it does sort of imply needing to either use the struct initializer syntax with zeroed out fields that will never be used, or adding more tightly-scoped initializer methods (a la `BytesSent::from_protocol`) if we wanted to hide that aspect. Neither make me feel warm and fuzzy inside.

Reply:
As you noted, we could use the same struct for both uses, and just zero out the values for registration. That's completely workable, and I could go for that approach. As you say, neither approach makes me feel warm and fuzzy.

I'd be tempted to also introduce a `struct Protocol<'a>(&'a str)` newtype to allow `register!(BytesSent::from(Protocol("grpc")))`, naming the types like we name the handle's data.

Incidentally, as it stands now, that would have to be `Registered<BytesSent<'static>>` due to the lifetime bound on `BytesSent`. That could be eliminated by making `protocol` in `BytesSent` be a `Cow` as well and drop the lifetime.

Reply:
If you're game for it, I'd say let's do everything you just said. The more we can wrap in newtypes, and wrap with `Cow`, and have be static, and so on and so forth... the more performant and misuse-resistant this stuff becomes.